packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts

```diff
@@ -0,0 +1,323 @@
+import { getUrlAndHeadersWithAuth } from "./getUrlAndHeadersWithAuth";
+import { getWithRetryForTokenRefresh } from "./odspUtils";
+import { TokenFetchOptions } from "./tokenFetch";
+
+/* eslint-disable max-len */
+
+interface IDedupCaches {
+    blobShaToPath: Map<string, string>,
+    pathToBlobSha: Map<string, string>,
+    treesPathToTree: Map<string, api.ISummaryTree>,
+}
+
+export class OdspSummaryUploadManager {
+    // This cache is associated with mapping sha to path for previous summary which belongs to last summary handle.
+    private blobTreeDedupCaches: IDedupCaches = {
+        blobShaToPath: new Map(),
+        pathToBlobSha: new Map(),
+        treesPathToTree: new Map(),
+    };
+
+    private backupBlobTreeDedupCaches: IDedupCaches = {
+        blobShaToPath: new Map(),
+        pathToBlobSha: new Map(),
+        treesPathToTree: new Map(),
+    };
+
+    // Last proposed handle of the uploaded app summary.
+    private lastSummaryProposalHandle: string | undefined;
+
+    constructor(
+        private readonly snapshotUrl: string,
+        private readonly getStorageToken: (options: TokenFetchOptions, name?: string) => Promise<string | null>,
+        private readonly logger: ITelemetryLogger,
+        private readonly epochTracker: EpochTracker,
+        private readonly blobCache: Map<string, IBlob | ArrayBuffer> = new Map(),
+    ) {
+    }
+
+    public async buildCachesForDedup(snapshotTree: api.ISnapshotTree, path: string = ""): Promise<api.ISummaryTree> {
+        assert(Object.keys(snapshotTree.commits).length === 0, "There should not be commit tree entries in snapshot");
+
+        const summaryTree: api.ISummaryTree = {
+            type: api.SummaryType.Tree,
+            tree: {},
+        };
+        for (const [key, value] of Object.entries(snapshotTree.blobs)) {
+            const blobValue = this.blobCache.get(value);
+            assert(blobValue !== undefined, "Blob should exists");
+            const decodedBlobValue = {
+                content: blobValue instanceof ArrayBuffer ? IsoBuffer.from(blobValue).toString("utf8") : fromBase64ToUtf8(blobValue.content),
+                encoding: "utf-8",
+            };
+            const hash = await hashFile(IsoBuffer.from(decodedBlobValue.content, decodedBlobValue.encoding));
+            // We are just setting the hash as content but we won't use it anywhere as it doesn't represent the actual content of blob.
+            summaryTree.tree[key] = {
+                type: api.SummaryType.Blob,
+                content: hash,
+            };
+            const fullBlobPath = path === "" ? `.app/${key}` : `${path}/${key}`;
+            this.blobTreeDedupCaches.blobShaToPath.set(hash, fullBlobPath);
+            this.blobTreeDedupCaches.pathToBlobSha.set(fullBlobPath, hash);
+        }
+
+        for (const [key, tree] of Object.entries(snapshotTree.trees)) {
+            const fullTreePath = path === "" ? key === ".protocol" ? `${key}` : ".app" : `${path}/${key}`;
```

Comment:
I'd put braces around double ? operator.
Also, why we ignore key in case of path === "" && key !== ".protocol"?
Is there any way for this code not to need to know about .app & .protocol ? I.e. shouldn't it be just a tree for this layer without any knowledge about meaning of special nodes? What if we have a 3rd entry in the list in the future?

Reply:
At some place we need to know about the ".app" and ".protocol". This is because of the structure of the snapshot tree today where for trees/blobs that belongs to ".protocol" starts with ".protocol" as parent node while the trees/blobs that belongs to app(datastores/dds), there path does not start with ".app". So we need to add ".app" somewhere. However with your suggestion I have localized it as much as possible. 
For future, we may not expect to add a lot of specialized entries like ".protocol", if we do then we may need to change the structure of snapshot tree or have a switch statement.

Reply:
Added braces and fixed key.

Reply:
Correct me if I got it wrong, but I think what you are saying is that upload and download formats for snapshot differ in the way they represent app tree. If it was not the case, then it would not matter for code like that - it would be just a tree, whiteout a need to understand specifics.

As such, can we confine that different to smallest part of our code base? I.e. not spread it across different parts of code? Or maybe create more obvious for understanding conversion functions that convert from one format to another, but their name clearly indicate why they are called, and what pre- and post- conditions are valid?

It's just extremely hard to follow all the logic (in general, this comment is not related to your PR :)), and someone joining later and not knowing specifics (or having clues to history) would be completely lost...

Reply:
So the downloaded tree is an ITree in which for detach new case all the paths are full starting with .app and .protocol however for other summaries where the ".app" is present as a commit, the path does not start with ".app" as the commit is app commit and server expects  us to deduce that which also makes sense. The combined protocol and app tree looks like this(which is understood by runtime):
tree = {
trees: {
".protocol": ....
"default":...
"scheduler":...
}
blobs: {
 // All blobs here belongs to app tree but there path does not start with ".app" as I think that is specific of odsp.
}
}
So we need to add ".app" in front of all entries which does not belong to protocol tree. Also the uploaded tree also has the same format where the entries does not container ".app" as prefix and we don't upload protocol tree at all.
I have localized this distinction only to this function which builds the caches.

Reply:
Can normalization (i.e. conversion to just one format) happen much higher in the stack? I'd hope that we return data in one format to runtime, not let every piece of code need to deal with two types of snapshots. It feels like here is a single choke point where such knowledge should be confined.

Reply:
Runtime always deals with the data in one format which is this:
    private combineProtocolAndAppSnapshotTree(
        snapshotTreeId: string,
        hierarchicalAppTree: api.ISnapshotTree,
        hierarchicalProtocolTree: api.ISnapshotTree,
    ) {
        const summarySnapshotTree: api.ISnapshotTree = {
            blobs: {
                ...hierarchicalAppTree.blobs,
            },
            commits: {
                ...hierarchicalAppTree.commits,
            },
            id: snapshotTreeId,
            trees: {
                ".protocol": hierarchicalProtocolTree,
                ...hierarchicalAppTree.trees,
            },
        };

        return summarySnapshotTree;
    }
This happens in the end and we populate our dedup caches with this format only.

Reply:
What I can do is have another api which will add ".app" in this trree at top level and then bnuild the caches. That way we will be able to remove ".app" from cache built function.

Reply:
I think that's the right direction. At least it makes code much more explicit / telling on what it does and why

Reply:
done